source/Timer.h

```diff
@@ -0,0 +1,121 @@
+// This requires a specification of how long to wait (base + randint(rand)),
+// a system to wait in, and whether to reset the timer upon leaving the system
+// and can optionally require the player to be idle (no acceleration), and/or
+// within or beyond a certain proximity to the system center or a
+// specific (named) system object
+class Timer {
+public:
+	Timer() = default;
+	Timer(const DataNode &node, const Mission *mission);
+	// Set up the timer from its data file node
+	void Load(const DataNode &node, const Mission *mission);
+	// Note: the Save() function can assume this is an instantiated Timer, not a template,
+	// so the time to wait will be saved fully calculated, and with any elapsed time subtracted
+	void Save(DataWriter &out) const;
+
+	// Get whether the timer is optional to complete
+	bool IsOptional() const;
+	// Get whether the timer is currently active
+	bool IsComplete() const;
+
+	// Calculate the total time to wait, including any random value
+	Timer Instantiate(std::map<std::string, std::string> &subs,
+					const System *origin, int jumps, int64_t payload) const;
+	// Progress the timer within the main loop
+	void Step(PlayerInfo &player, UI *ui);
+
+private:
+	enum class ResetCondition {
+		NONE,
+		PAUSE,
+		LEAVE_ZONE,
+		LEAVE_SYSTEM
+	};
+	void ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui);
+
+private:
+	// The name of the timer
+	std::string name;
+	// The basic amount of time to wait, with the optional random values
+	uint64_t base = 0;
+	uint32_t rand = 0;
+
+	// The system the timer is for
+	const System *system = nullptr;
+	// ...Or the systems it can be for
+	LocationFilter systems;
+
+	// If set, the mission the timer is attached to can be completed even without it
+	bool optional = false;
+	// Whether the timer requires the player to be idle
+	bool requireIdle = false;
+	// Whether the timer requires the player to be uncloaked to advance
+	bool requireUncloaked = false;
+	// What circumstances will reset the timer: leaving the system,
+	// leaving the proximity zone (if applicable), or any circumstance that stops the timer
+	ResetCondition resetCondition = Timer::ResetCondition::PAUSE;
+	// If proximity is specified, this determines whether the timer will only advance while
+	// close to or far from the specified center; default is close to
+	bool closeTo = true;
+	// This specifies the radius around the proximity center to be checked
+	double proximity = 0.;
+	// This specifies the system object to be the center; if this is nullptr and the filter is empty,
+	// but proximity is positive, then the system center is used instead
+	const Planet *proximityCenter = nullptr;
+	// The filter to use for which planet(s) to hang out near if a single planet isn't set
+	LocationFilter proximityCenters;
+
+	// The action to be performed when the timer elapses
+	MissionAction action;
+	// The action to be performed when the timer resets
+	MissionAction resetAction;
```

Comment:
Hm. Yes, I think that makes sense—particularly since I've already thought that we _might_ want an "on timer begin" action (but haven't added it because it's easy to simulate by adding a second timer with the same parameters but a much shorter time).

Is it better in a case like this to overload the `Trigger` enum name (which I _think_ I understand should still be valid, since it would be within the Timer class...?), or to make a new name like `TimerTrigger` for full clarity?

Reply:
> Is it better in a case like this to overload the Trigger enum name, or to make a new name like TimerTrigger for full clarity?

Honestly, either way is fine. We already have Trigger defined in both NPC and Mission, but it's not like it hurts to be specific.